### PR TITLE
LibCompress: Implement Brotli decompressor

### PR DESCRIPTION
**Pull request by msvisser**
This pull request adds a Brotli decompressor implementation.

The current version of this pull request is functional, but needs a lot of clean-up and refactoring, therefore it is a draft.

### CODE REVIEW

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 1]
```diff
@@ -0,0 +1,1149 @@
...
+
+#include <AK/BinarySearch.h>
+#include <AK/MemoryStream.h>
+#include <AK/QuickSort.h>
+#include <LibCompress/Brotli.h>
+#include <LibCompress/BrotliDictionary.h>
+
+namespace Compress {
+
+Optional<size_t> BrotliCanonicalCode::read_symbol(InputBitStream& input_bit_stream)
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCompress/Brotli.cpp:15]: I'm guessing that the input bit stream is little endian? Then it would be great to not have an old stream for new code, but use a reimplementation `LittleEndianInputBitStream` with `Core::Stream`. This will be super convenient for porting other bit stream users!

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 2]
```diff
@@ -0,0 +1,1149 @@
...
+
+    const auto nread = min(m_bytes_remaining, m_decompressor.m_output_stream->remaining_contiguous_space());
+    m_bytes_remaining -= nread;
+
+    m_decompressor.m_input_stream >> m_decompressor.m_output_stream->reserve_contiguous_space(nread);
+
+    return true;
+}
+
+BrotliDecompressor::BrotliDecompressor(InputStream& input_stream)
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCompress/Brotli.cpp:398]: As above, can you use a `Core::Stream::InputStream` instead?

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 3]
```diff
@@ -0,0 +1,1149 @@
...
+            if (m_input_stream.has_any_error()) {
+                set_fatal_error();
+                break;
+            }
+
+            total_read += nread;
+            if (nread == slice.size())
+                break;
+
+            m_compressed_block.~CompressedBlock();
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCompress/Brotli.cpp:1060]: Why is this necessary?

File: Userland/Libraries/LibCompress/Brotli.h [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+public:
+    BrotliCanonicalCode() = default;
+    Optional<size_t> read_symbol(InputBitStream&);
+
+private:
+    Vector<size_t> m_symbol_codes;
+    Vector<size_t> m_symbol_values;
+};
+
+class BrotliDecompressor : public InputStream {
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCompress/Brotli.h:29]: `Core::Stream::InputStream` (just reminder)

File: Userland/Libraries/LibCompress/BrotliDictionary.cpp [hunk 5]
```diff
@@ -0,0 +1,7908 @@
...
+ */
+
+#include <AK/Format.h>
+#include <AK/Types.h>
+#include <LibCompress/BrotliDictionary.h>
+
+namespace Compress {
+
+static u8 dictionary_data[122784] {
+    0x74, 0x69, 0x6d, 0x65, 0x64, 0x6f, 0x77, 0x6e, 0x6c, 0x69, 0x66, 0x65, 0x6c, 0x65, 0x66, 0x74,
```
Comment (creator1creeper1) [inline Userland/Libraries/LibCompress/BrotliDictionary.cpp:14]: Can this be stored in a binary format somehow? It feels pretty wasteful to have it in source code.

**Review by creator1creeper1 — COMMENTED** [review]
Impressive.

**Comment by msvisser** [comment]
@kleinesfilmroellchen I will refactor this code to work with the `Core::Stream` API, which seems much nicer, especially for error handling. I will also have a look at moving all the test files to some external location, although I'm not sure how that would work correctly in both the Lagom and on-system tests.

@creator1creeper1 Yeah it should probably be stored in a separate file. I know that Rust has the `include_bytes!` macro, which allows you to include a binary file as a byte array, however I don't know the C++ equivalent way of doing that.

**Comment by kleinesfilmroellchen** [comment]
I'm sorry, this is a bit over my head. It should somehow be possible to move the files into a known location as part of the build system, and then use normal file operations to read them.

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 6]
```diff
@@ -0,0 +1,905 @@
...
+ErrorOr<size_t> BrotliDecompressionStream::read_window_length()
+{
+    if (TRY(m_input_stream.read_bit())) {
+        switch (TRY(m_input_stream.read_bits(3))) {
+        case 0: {
+            switch (TRY(m_input_stream.read_bits(3))) {
+            case 0:
+                return 17;
+            case 1:
+                VERIFY_NOT_REACHED();
```
Comment (Lubrsi) [inline Userland/Libraries/LibCompress/Brotli.cpp:45]: This can be reached with this test file:
[crash-5fee003afd41dd02b9a21df2ae7f9749c57f7404.txt](https://github.com/SerenityOS/serenity/files/8421858/crash-5fee003afd41dd02b9a21df2ae7f9749c57f7404.txt)

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 7]
```diff
@@ -0,0 +1,905 @@
...
+
+        if (len != 0) {
+            code_length_counts[len]++;
+            sum += (32 >> len);
+            number_of_non_zero_symbols++;
+        }
+
+        if (sum == 32)
+            break;
+        VERIFY(sum < 32);
```
Comment (Lubrsi) [inline Userland/Libraries/LibCompress/Brotli.cpp:256]: This can be reached with this test file:
[crash-8e13ebed311edf16682ac6004d77d07000bc2933.txt](https://github.com/SerenityOS/serenity/files/8421866/crash-8e13ebed311edf16682ac6004d77d07000bc2933.txt)

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 8]
```diff
@@ -0,0 +1,905 @@
...
+    }
+    // dbgln("run_length_encoding_max: {}", run_length_encoding_max);
+
+    BrotliDecompressionStream::CanonicalCode code;
+    TRY(read_prefix_code(code, number_of_codes + run_length_encoding_max));
+
+    size_t i = 0;
+    while (i < context_map_size) {
+        // dbgln("i: {}", i);
+        size_t symbol = code.read_symbol(m_input_stream).release_value();
```
Comment (Lubrsi) [inline Userland/Libraries/LibCompress/Brotli.cpp:413]:
```suggestion
        size_t symbol = TRY(code.read_symbol(m_input_stream));
```
Test file:
[crash-0834143266645039e9ceb21619ffa94920ceb4cb.txt](https://github.com/SerenityOS/serenity/files/8421893/crash-0834143266645039e9ceb21619ffa94920ceb4cb.txt)

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 9]
```diff
@@ -0,0 +1,905 @@
...
+
+ErrorOr<void> BrotliDecompressionStream::block_update_length(Block& block)
+{
+    size_t const block_length_code_base[26] { 1, 5, 9, 13, 17, 25, 33, 41, 49, 65, 81, 97, 113, 145, 177, 209, 241, 305, 369, 497, 753, 1265, 2289, 4337, 8433, 16625 };
+    size_t const block_length_code_extra[26] { 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 5, 6, 6, 7, 8, 9, 10, 11, 12, 13, 24 };
+
+    size_t symbol = TRY(block.length_code.read_symbol(m_input_stream));
+    // dbgln("block length symbol: {}", symbol);
+
+    size_t block_length = block_length_code_base[symbol] + TRY(m_input_stream.read_bits(block_length_code_extra[symbol]));
```
Comment (Lubrsi) [inline Userland/Libraries/LibCompress/Brotli.cpp:476]: These buffers can be overflowed:
[crash-f87d381e0dcfe2928983707d18780486c71c24d7.txt](https://github.com/SerenityOS/serenity/files/8421905/crash-f87d381e0dcfe2928983707d18780486c71c24d7.txt)
[crash-4daea4fa729099ae5db13a334e4f0e9a2bc5a4a6.txt](https://github.com/SerenityOS/serenity/files/8421941/crash-4daea4fa729099ae5db13a334e4f0e9a2bc5a4a6.txt)

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 10]
```diff
@@ -0,0 +1,905 @@
...
+            result_lengths_count[symbol]++;
+
+            i++;
+
+            if (symbol != 0) {
+                previous_non_zero_code_length = symbol;
+                sum += (32768 >> symbol);
+                if (sum == 32768)
+                    break;
+                VERIFY(sum < 32768);
```
Comment (Lubrsi) [inline Userland/Libraries/LibCompress/Brotli.cpp:313]: This can be reached with: [crash-f387b948147677a78d1709903863eac40ab88a78.txt](https://github.com/SerenityOS/serenity/files/8421915/crash-f387b948147677a78d1709903863eac40ab88a78.txt)

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 11]
```diff
@@ -0,0 +1,905 @@
...
+    size_t previous_non_zero_code_length = 8;
+    size_t last_symbol = 0;
+    size_t last_repeat = 0;
+
+    Vector<size_t> result_symbols;
+    Vector<size_t> result_lengths;
+    size_t result_lengths_count[16] { 0 };
+    while (i < alphabet_size) {
+        VERIFY(sum < 32768);
+        auto symbol = temp_code.read_symbol(m_input_stream).release_value();
```
Comment (Lubrsi) [inline Userland/Libraries/LibCompress/Brotli.cpp:299]:
```suggestion
        auto symbol = TRY(temp_code.read_symbol(m_input_stream));
```

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 12]
```diff
@@ -0,0 +1,905 @@
...
+ErrorOr<size_t> BrotliDecompressionStream::CanonicalCode::read_symbol(LittleEndianInputBitStream& input_stream)
+{
+    size_t code_bits = 1;
+
+    while (code_bits < (1 << 16)) {
+        // FIXME: This is very inefficient and could greatly be improved by implementing this
+        //        algorithm: https://www.hanshq.net/zip.html#huffdec
+        size_t index;
+        if (binary_search(m_symbol_codes.span(), code_bits, &index))
+            return m_symbol_values[index];
```
Comment (Lubrsi) [inline Userland/Libraries/LibCompress/Brotli.cpp:23]: This index can be outside the vector, I get this when going to `cloudflare.com` sometimes, but do not have an example file on hand.
```
12.509 RequestServer(39:39): ASSERTION FAILED: i < m_size
../.././AK/Vector.h:149
12.509 [#0 RequestServer(39:39)]: Terminating RequestServer(39) due to signal 6
12.517 [#0 FinalizerTask(4:4)]: Backtrace:
0x00000000deadc0de  Kernel::Processor::switch_context(Kernel::Thread*&, Kernel::Thread*&) + 0x383
12.517 [#0 FinalizerTask(4:4)]: Generating coredump for pid: 39
12.552 CrashDaemon(9:9): New coredump file: /tmp/coredump/RequestServer_39_1649194427
12.659 WebContent(42:42): Parse error! [handle_initial @ ../../Userland/Libraries/LibWeb/HTML/Parser/HTMLParser.cpp:469]
12.713 CrashReporter(41:41): Started thread "", tid = 43
13.687 CrashReporter(41:43): Generating backtrace took 969 ms
13.687 CrashReporter(41:43): --- Backtrace for thread #0 (TID 39) ---
13.687 CrashReporter(41:43): 0x000000028d35a02b: [libsystem.so] syscall2 +0xb (syscall.cpp:25 => syscall.cpp:24)
13.687 CrashReporter(41:43): 0x0000000b0999eba7: [libc.so] abort +0x27 (stdlib.cpp:223)
13.687 CrashReporter(41:43): 0x0000000b099a3b3e: [libc.so] __assertion_failed +0x7e (assert.cpp:33)
13.690 CrashReporter(41:43): 0x000000166c084117: [libcompress.so.serenity] Compress::BrotliDecompressionStream::CanonicalCode::read_symbol(Core::Stream::LittleEndianInputBitStream&) [clone .localalias] +0x197 (Vector.h:149 => Vector.h:157 => Brotli.cpp:23)
13.690 CrashReporter(41:43): 0x000000166c086904: [libcompress.so.serenity] Compress::BrotliDecompressionStream::block_update_length(Compress::BrotliDecompressionStream::Block&) +0x264 (Brotli.cpp:473)
13.690 CrashReporter(41:43): 0x000000166c086c52: [libcompress.so.serenity] Compress::BrotliDecompressionStream::read_block_configuration(Compress::BrotliDecompressionStream::Block&) +0xf2 (Brotli.cpp:462)
13.690 CrashReporter(41:43): 0x000000166c087d3c: [libcompress.so.serenity] Compress::BrotliDecompressionStream::read(AK::Span<unsigned char>) +0x107c (Brotli.cpp:647)
13.690 CrashReporter(41:43): 0x0000000f66613289: [libcore.so.serenity] Core::Stream::Stream::read_all() +0x69 (Stream.cpp:57)
13.695 CrashReporter(41:43): 0x000000109b68a0f4: [libhttp.so.serenity] HTTP::Job::finish_up() [clone .localalias] +0xff4 (Job.cpp:86 => Job.cpp:613)
13.695 CrashReporter(41:43): 0x000000109b68ce4f: [libhttp.so.serenity] HTTP::Job::on_socket_connected()::{lambda()#2}::operator()() const +0x251f (Job.cpp:323)
13.695 CrashReporter(41:43): 0x000000109b6875c3: [libhttp.so.serenity] AK::Function<void ()>::CallableWrapper<HTTP::Job::register_on_ready_to_read(AK::Function<void ()>)::{lambda()#1}>::call() +0x63 (Function.h:91)
13.695 CrashReporter(41:43): 0x000000109b687880: [libhttp.so.serenity] AK::Function<void ()>::CallableWrapper<HTTP::Job::register_on_ready_to_read(AK::Function<void ()>)::{lambda()#1}::operator()() const::{lambda()#2}>::call() +0x50 (Function.h:91)
13.699 CrashReporter(41:43): 0x0000000f6660880c: [libcore.so.serenity] AK::Function<void ()>::CallableWrapper<Core::Object::deferred_invoke(AK::Function<void ()>)::{lambda()#1}>::call() +0x23c (Function.h:91)
13.699 CrashReporter(41:43): 0x0000000f665f1a6a: [libcore.so.serenity] Core::EventLoop::pump(Core::EventLoop::WaitMode) [clone .localalias] +0x28a (Function.h:91)
13.699 CrashReporter(41:43): 0x0000000f665f2229: [libcore.so.serenity] Core::EventLoop::exec() +0x119 (EventLoop.cpp:438)
13.699 CrashReporter(41:43): 0x0000001a965a05c1: [/bin/RequestServer] serenity_main(Main::Arguments) +0x1b1 (main.cpp:51)
13.699 CrashReporter(41:43): 0x0000001a96573915: [/bin/RequestServer] main +0x95 (Main.cpp:39)
13.702 CrashReporter(41:43): 0x0000001a96573b5f: [/bin/RequestServer] _entry +0x6f (crt0.cpp:49)
```

File: Userland/Libraries/LibCompress/BrotliDictionary.cpp [hunk 13]
```diff
@@ -0,0 +1,7902 @@
...
+    0x63, 0x6f, 0x6c, 0x6f, 0x6d, 0x62, 0x69, 0x61, 0x64, 0x65, 0x73, 0x70, 0x75, 0xc3, 0xa9, 0x73,
+    0x64, 0x65, 0x70, 0x6f, 0x72, 0x74, 0x65, 0x73, 0x70, 0x72, 0x6f, 0x79, 0x65, 0x63, 0x74, 0x6f,
+    0x70, 0x72, 0x6f, 0x64, 0x75, 0x63, 0x74, 0x6f, 0x70, 0xc3, 0xba, 0x62, 0x6c, 0x69, 0x63, 0x6f,
+    0x6e, 0x6f, 0x73, 0x6f, 0x74, 0x72, 0x6f, 0x73, 0x68, 0x69, 0x73, 0x74, 0x6f, 0x72, 0x69, 0x61,
+    0x70, 0x72, 0x65, 0x73, 0x65, 0x6e, 0x74, 0x65, 0x6d, 0x69, 0x6c, 0x6c, 0x6f, 0x6e, 0x65, 0x73,
+    0x6d, 0x65, 0x64, 0x69, 0x61, 0x6e, 0x74, 0x65, 0x70, 0x72, 0x65, 0x67, 0x75, 0x6e, 0x74, 0x61,
+    0x61, 0x6e, 0x74, 0x65, 0x72, 0x69, 0x6f, 0x72, 0x72, 0x65, 0x63, 0x75, 0x72, 0x73, 0x6f, 0x73,
+    0x70, 0x72, 0x6f, 0x62, 0x6c, 0x65, 0x6d, 0x61, 0x73, 0x61, 0x6e, 0x74, 0x69, 0x61, 0x67, 0x6f,
+    0x6e, 0x75, 0x65, 0x73, 0x74, 0x72, 0x6f, 0x73, 0x6f, 0x70, 0x69, 0x6e, 0x69, 0xc3, 0xb3, 0x6e,
+    0x69, 0x6d, 0x70, 0x72, 0x69, 0x6d, 0x69, 0x72, 0x6d, 0x69, 0x65, 0x6e, 0x74
```
Comment (Lubrsi) [inline Userland/Libraries/LibCompress/BrotliDictionary.cpp:7866]: This can fail, for example by going to `cloudflare.com`:
```
28.845 RequestServer(44:44): ASSERTION FAILED: 4 <= length && length <= 24
../../Userland/Libraries/LibCompress/BrotliDictionary.cpp:7866
28.848 [#0 RequestServer(44:44)]: Terminating RequestServer(44) due to signal 6
28.848 [#0 FinalizerTask(4:4)]: Backtrace:
0x00000000deadc0de  Kernel::Processor::switch_context(Kernel::Thread*&, Kernel::Thread*&) + 0x383
28.848 [#0 FinalizerTask(4:4)]: Generating coredump for pid: 44
28.879 CrashDaemon(9:9): New coredump file: /tmp/coredump/RequestServer_44_1649194443
29.005 WebContent(46:46): Parse error! [handle_initial @ ../../Userland/Libraries/LibWeb/HTML/Parser/HTMLParser.cpp:469]
29.013 CrashReporter(45:45): Started thread "", tid = 47
29.382 CrashReporter(45:47): Generating backtrace took 357 ms
29.382 CrashReporter(45:47): --- Backtrace for thread #0 (TID 44) ---
29.382 CrashReporter(45:47): 0x0000001b995f702b: [libsystem.so] syscall2 +0xb (syscall.cpp:25 => syscall.cpp:24)
29.382 CrashReporter(45:47): 0x0000000a57877ba7: [libc.so] abort +0x27 (stdlib.cpp:223)
29.382 CrashReporter(45:47): 0x0000000a5787cb3e: [libc.so] __assertion_failed +0x7e (assert.cpp:33)
29.382 CrashReporter(45:47): 0x0000000460cc795b: [libcompress.so.serenity] Compress::BrotliDictionary::lookup_word(unsigned long, unsigned long) +0x20b (BrotliDictionary.cpp:7866)
29.382 CrashReporter(45:47): 0x0000000460cc52b3: [libcompress.so.serenity] Compress::BrotliDecompressionStream::read(AK::Span<unsigned char>) +0x15f3 (Brotli.cpp:853)
29.386 CrashReporter(45:47): 0x000000181b0ec289: [libcore.so.serenity] Core::Stream::Stream::read_all() +0x69 (Stream.cpp:57)
29.386 CrashReporter(45:47): 0x000000183ddcc0f4: [libhttp.so.serenity] HTTP::Job::finish_up() [clone .localalias] +0xff4 (Job.cpp:86 => Job.cpp:613)
29.386 CrashReporter(45:47): 0x000000183ddcee4f: [libhttp.so.serenity] HTTP::Job::on_socket_connected()::{lambda()#2}::operator()() const +0x251f (Job.cpp:323)
29.386 CrashReporter(45:47): 0x000000183ddc95c3: [libhttp.so.serenity] AK::Function<void ()>::CallableWrapper<HTTP::Job::register_on_ready_to_read(AK::Function<void ()>)::{lambda()#1}>::call() +0x63 (Function.h:91)
29.386 CrashReporter(45:47): 0x000000183ddc9880: [libhttp.so.serenity] AK::Function<void ()>::CallableWrapper<HTTP::Job::register_on_ready_to_read(AK::Function<void ()>)::{lambda()#1}::operator()() const::{lambda()#2}>::call() +0x50 (Function.h:91)
29.390 CrashReporter(45:47): 0x000000181b0e180c: [libcore.so.serenity] AK::Function<void ()>::CallableWrapper<Core::Object::deferred_invoke(AK::Function<void ()>)::{lambda()#1}>::call() +0x23c (Function.h:91)
29.390 CrashReporter(45:47): 0x000000181b0caa6a: [libcore.so.serenity] Core::EventLoop::pump(Core::EventLoop::WaitMode) [clone .localalias] +0x28a (Function.h:91)
29.390 CrashReporter(45:47): 0x000000181b0cb229: [libcore.so.serenity] Core::EventLoop::exec() +0x119 (EventLoop.cpp:438)
29.390 CrashReporter(45:47): 0x00000006bff5b5c1: [/bin/RequestServer] serenity_main(Main::Arguments) +0x1b1 (main.cpp:51)
29.390 CrashReporter(45:47): 0x00000006bff2e915: [/bin/RequestServer] main +0x95 (Main.cpp:39)
29.394 CrashReporter(45:47): 0x00000006bff2eb5f: [/bin/RequestServer] _entry +0x6f (crt0.cpp:49)
```
I got a fuzzed example file:
[crash-cfea2e3d833661611ae605b1a62e68a23d110dae.txt](https://github.com/SerenityOS/serenity/files/8422042/crash-cfea2e3d833661611ae605b1a62e68a23d110dae.txt)

Comment (Lubrsi) [inline Userland/Libraries/LibCompress/BrotliDictionary.cpp:7872]: This can fail:
[crash-8fe6d7ff55d1c699f3e8f979201e32e90389da43.txt](https://github.com/SerenityOS/serenity/files/8422021/crash-8fe6d7ff55d1c699f3e8f979201e32e90389da43.txt)

**Comment by Lubrsi** [comment]
This is really cool! :^)
I wrote a fuzzer for this to find some issues:
```cpp
/*
 * Copyright (c) 2022, the SerenityOS developers.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <LibCore/MemoryStream.h>
#include <LibCompress/Brotli.h>
#include <stdio.h>

extern "C" int LLVMFuzzerTestOneInput(uint8_t const* data, size_t size)
{
    auto bufstream_result = Core::Stream::MemoryStream::construct({ const_cast<uint8_t*>(data), size });
    if (bufstream_result.is_error()) {
        dbgln("Job::handle_content_encoding: MemoryStream::construct() failed.");
        return 1;
    }
    auto bufstream = bufstream_result.release_value();

    auto brotli_stream = Compress::BrotliDecompressionStream { *bufstream };

    auto uncompressed = brotli_stream.read_all();
    return uncompressed.is_error();
}
```
```patch
diff --git a/Meta/Lagom/Fuzzers/CMakeLists.txt b/Meta/Lagom/Fuzzers/CMakeLists.txt
index 6fa8da4750..23ca4a48bb 100644
--- a/Meta/Lagom/Fuzzers/CMakeLists.txt
+++ b/Meta/Lagom/Fuzzers/CMakeLists.txt
@@ -16,6 +16,7 @@ function(add_simple_fuzzer name)
 endfunction()
 
 add_simple_fuzzer(FuzzBMPLoader LagomGfx)
+add_simple_fuzzer(FuzzBrotli LagomCompress)
 add_simple_fuzzer(FuzzCyrillicDecoder LagomTextCodec)
 add_simple_fuzzer(FuzzDeflateCompression LagomCompress)
 add_simple_fuzzer(FuzzDeflateDecompression LagomCompress)
 ```
 These are the results! I've uploaded crash files from the fuzzer with a `.txt` extension so GitHub accepts them, but they are all really raw binary files.

**Review by Lubrsi — CHANGES_REQUESTED** [review]
This is really cool! :^)
I wrote a fuzzer for this to find some issues:
```cpp
/*
 * Copyright (c) 2022, the SerenityOS developers.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <LibCore/MemoryStream.h>
#include <LibCompress/Brotli.h>
#include <stdio.h>

extern "C" int LLVMFuzzerTestOneInput(uint8_t const* data, size_t size)
{
    auto bufstream_result = Core::Stream::MemoryStream::construct({ const_cast<uint8_t*>(data), size });
    if (bufstream_result.is_error()) {
        dbgln("Job::handle_content_encoding: MemoryStream::construct() failed.");
        return 1;
    }
    auto bufstream = bufstream_result.release_value();

    auto brotli_stream = Compress::BrotliDecompressionStream { *bufstream };

    auto uncompressed = brotli_stream.read_all();
    return uncompressed.is_error();
}
```
```patch
diff --git a/Meta/Lagom/Fuzzers/CMakeLists.txt b/Meta/Lagom/Fuzzers/CMakeLists.txt
index 6fa8da4750..23ca4a48bb 100644
--- a/Meta/Lagom/Fuzzers/CMakeLists.txt
+++ b/Meta/Lagom/Fuzzers/CMakeLists.txt
@@ -16,6 +16,7 @@ function(add_simple_fuzzer name)
 endfunction()
 
 add_simple_fuzzer(FuzzBMPLoader LagomGfx)
+add_simple_fuzzer(FuzzBrotli LagomCompress)
 add_simple_fuzzer(FuzzCyrillicDecoder LagomTextCodec)
 add_simple_fuzzer(FuzzDeflateCompression LagomCompress)
 add_simple_fuzzer(FuzzDeflateDecompression LagomCompress)
 ```
 These are the results! I've uploaded crash files from the fuzzer with a `.txt` extension so GitHub accepts them, but they are all really raw binary files.

**Comment by msvisser** [comment]
Thanks for the large list of fuzzing results.
Most of those assertions should actually be error handling, but lazy me put in the assertions during development.
I already found some other problems with this fuzzer, which I will fix too!

**Comment by msvisser** [comment]
With these changes I think all of the fuzzer bugs are fixed, please let me know if you find any more.
There is still some logical error I think, since https://www.cloudflare.com/ still does not load correctly.
Also there might be a bug in LibHTTP/Job, which causes a nullptr deref on decompression error, however I haven't had the time to locate the exact problem.

**Comment by msvisser** [comment]
These last changes should fix the problem with https://cloudflare.com.
I was not clearing out the state of the decompressor when starting a new meta-block, which resulted in using the wrong prefix codes for the new block.

The testcase that I added does seem to annoy the linter, as the file does not end in a newline. Unfortunately, I cannot reproduce the exact compression of this file, so I cannot compress the file with a newline. Is there a way to make the linter ignore all files in the test files folder, as they are not really source code?

File: Tests/LibCompress/TestBrotli.cpp [hunk 14]
```diff
@@ -75,3 +75,8 @@ TEST_CASE(brotli_decompress_duckduckgo_css2)
 {
     run_test("s2049.css");
 }
+
+TEST_CASE(brotli_decompress_cloudflare_html)
+{
+    run_test("cloudflare.html");
```
Comment (kleinesfilmroellchen) [inline Tests/LibCompress/TestBrotli.cpp:81]: You've extracted this nicely, but is cloudflare's HTML something we can legally include? I don't think so. Let's either use a website that's explicitly BSD compatible, or one of our own static sites, like a birthday announcement post.

Comment (msvisser) [reply Tests/LibCompress/TestBrotli.cpp:81]: Yeah it might be a bit tricky, there are also some CSS files from DuckDuckGo in the tests too.

The problem is that this specific compressed file caused a problem with the implementation, while I could not reproduce it with other files.

In the end might just remove these tests, or replace them with actual HTTP requests, so the files are not included in our repository.

Comment (kleinesfilmroellchen) [reply Tests/LibCompress/TestBrotli.cpp:81]: How about you minimize the cloudflare problem and create an artificial test that's far enough from their HTML it doesn't cause problems?

Regarding DuckDuckGo, that also looks like a problem even though they would probably care less.

Comment (msvisser) [reply Tests/LibCompress/TestBrotli.cpp]: I have removed all of the test cases that might be a possible copyright problem.
Instead I'm using a few of the SerentiyOS website pages, the Katica font file, and a tar of the Userland source code.

**Comment by msvisser** [comment]
I have rebased this pull request to squash a bunch of commits.
Should be good to go now, unless there are any other issues.

Only thing is that we might want to move the BrotliDictionary data to a separate file, however I'm uncertain how to do that in a nice way.

File: Tests/LibCompress/TestBrotli.cpp [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+#ifdef __serenity__
+    String path = String::formatted("/usr/Tests/LibCompress/brotli-test-files/{}", file_name);
+#else
+    String path = String::formatted("../../Tests/LibCompress/brotli-test-files/{}", file_name);
+#endif
+
+    auto cmp_file = MUST(Core::Stream::File::open(path, Core::Stream::OpenMode::Read));
+    auto cmp_data = MUST(cmp_file->read_all());
+
+    String path_compressed = String::formatted("{}.br", path);
```
Comment (kleinesfilmroellchen) [inline Tests/LibCompress/TestBrotli.cpp:25]: Why don't you need the weird path handling here? That looks suspicious.

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 16]
```diff
@@ -0,0 +1,920 @@
...
+            }
+        }
+    }
+
+    return {};
+}
+
+static void inverse_move_to_front_transform(Span<u8> v)
+{
+    // RFC 7932 section 7.3
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCompress/Brotli.cpp:391]: These could be elsewhere as well, probably? (I know FLAC also needs spec comments, I'll do that at some point...)

File: Userland/Libraries/LibCompress/Brotli.cpp [hunk 17]
```diff
@@ -0,0 +1,920 @@
...
+}
+
+ErrorOr<void> BrotliDecompressionStream::read_context_map(size_t number_of_codes, Vector<u8>& context_map, size_t context_map_size)
+{
+    bool use_run_length_encoding = TRY(m_input_stream.read_bit());
+    size_t run_length_encoding_max = 0;
+    if (use_run_length_encoding) {
+        run_length_encoding_max = 1 + TRY(m_input_stream.read_bits(4));
+    }
+    // dbgln("run_length_encoding_max: {}", run_length_encoding_max);
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCompress/Brotli.cpp:414]: Leftovers, also below

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]

File: Tests/LibCompress/TestBrotli.cpp (hunk 15 above)
Comment (msvisser) [reply Tests/LibCompress/TestBrotli.cpp:25]: This is just appending the `br` extension at the end of the file path.

File: Userland/Libraries/LibCompress/Brotli.cpp (hunk 16 above)
Comment (msvisser) [reply Userland/Libraries/LibCompress/Brotli.cpp:391]: Yeah, some more references to the RFC might be nice.

This specific function has a reference since it was defined in the specification.

Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibCompress/Brotli.cpp:391]: If it's an RFC, it's basically public domain.

File: Userland/Libraries/LibHTTP/Job.cpp [hunk 18]
```diff
@@ -66,6 +68,31 @@ static Optional<ByteBuffer> handle_content_encoding(const ByteBuffer& buf, const
             dbgln("  Output size: {}", uncompressed.value().size());
         }
 
+        return uncompressed.release_value();
+    } else if (content_encoding == "br") {
+        dbgln_if(JOB_DEBUG, "Job::handle_content_encoding: buf is brotli compressed!");
+
+        // FIXME: MemoryStream is both read and write, however we only need the read part here
```
Comment (alimpfard) [inline Userland/Libraries/LibHTTP/Job.cpp:75]: Could you also add a FIXME for streaming these instead of doing them all at once at the end?

**Review by alimpfard — COMMENTED** [review]
Perhaps we could generate the dictionary file via cmake from a binary representation?
8kloc of that is a bit too much :P

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by msvisser** [comment]
I have no clue why the CI build is failing, it seems to be breaking on code that is not actually in my version of the repo.

**Comment by msvisser** [comment]
It seems that compiling Lagom on macOS does not like my little assembly hack for including the dictionary file.
```C++
// Include the 119.9 KiB of dictionary data from a binary file
extern u8 const brotli_dictionary_data[];
asm(
    ".section \".rodata\"\n"
    "brotli_dictionary_data:\n"
    ".incbin \"LibCompress/BrotliDictionaryData.bin\"\n"
    ".previous\n");
```
Also, the tests don't seem to work on Azure Linux.

**Comment by kleinesfilmroellchen** [comment]
Squash commits?

**Comment by msvisser** [comment]
Commits are squashed down to what I think is reasonable.
Also, I fixed the tests so that CI is green now!
As far as I am concerned this is good to go.

**Review by kleinesfilmroellchen — APPROVED** [review]
